packages/design-system/src/components/dialog.tsx

```diff
@@ -21,23 +24,121 @@ export const DialogClose = Primitive.Close;
 // https://www.radix-ui.com/docs/primitives/components/dialog#description
 export const DialogDescription = Primitive.Description;
 
+const placeholderImage =
+  typeof Image !== "undefined" ? new Image(0, 0) : undefined;
+// It's important to set the src early, because it has to be loaded by the time drag starts.
+if (placeholderImage) {
+  placeholderImage.src =
+    "data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==";
+}
+
+type UseDraggableProps = {
+  isDraggable?: boolean;
+  isMaximized?: boolean;
+  width?: number;
+  height?: number;
+};
+
+const useDraggable = ({
+  isDraggable = false,
+  isMaximized = false,
+  width,
+  height,
+}: UseDraggableProps) => {
+  const initialDataRef = useRef<{
+    point: { x: number; y: number };
+    rect: DOMRect;
+  }>();
+  const { enableCanvasPointerEvents, disableCanvasPointerEvents } =
+    useDisableCanvasPointerEvents();
+
+  const handleDragStart: DragEventHandler = (event) => {
+    disableCanvasPointerEvents();
+    if (placeholderImage) {
+      event.dataTransfer.setDragImage(placeholderImage, 0, 0);
+    }
+    const target = event.target as HTMLElement;
+    const rect = target.getBoundingClientRect();
+    target.style.left = `${rect.left}px`;
+    target.style.top = `${rect.top}px`;
+    target.style.transform = "none";
+    initialDataRef.current = {
+      point: { x: event.pageX, y: event.pageY },
+      rect,
+    };
+  };
+
+  const handleDrag: DragEventHandler = (event) => {
+    event.preventDefault();
+    if (
+      event.pageX <= 0 ||
+      event.pageY <= 0 ||
+      initialDataRef.current === undefined
+    ) {
+      return;
+    }
+    const { rect, point } = initialDataRef.current;
+    const movementX = point.x - event.pageX;
+    const movementY = point.y - event.pageY;
+    let left = Math.max(rect.x - movementX, 0);
+    left = Math.min(left, window.innerWidth - rect.width);
+    let top = Math.max(rect.y - movementY, 0);
+    // We want some part of the dialog to be visible but otherwise let it go off screen.
```

Comment:
Strange comment starting from 'but', Im not getting it. 
IMO this `  // We want some part of the dialog to be visible` is enough

Reply:
Also why top only?

Reply:
intentional behavior so that the header of the dialog is always fully visible and so that you can make the dialog mostly offscreen and see whats behind